### PR TITLE
Support drawing arrows to show case relationship & filter cases on the map

### LINKED ISSUE #237: G3.10 Support drawing arrows between points based on time series information in host application
Description from user story ": I want to view the location of all cases in an area, with a visual representation (e.g., arrows, lines) of the relationship between each index case and introduced cases related to each index case."

We need to support the ability to draw arrows on the map from one point to the next showing how a case has moved from point 1 to 2 to 3 to 4, etc. The arrows should be drawn directly as lines, not following any types of path or road.

Use Case:
- A user moves from house to house testing each member for malaria.
- If the malaria test is positive, they provide treatment immediately and each test is recorded in the host application.
- Forms are submitted at every house as the user moves through the village and each form contains time series information (formSubmission datetime).
- The user wants to view the location of all cases in the area displayed on the GeoWidget so they can see how they moved throughout the village and all of the positive cases.

Development Required:
The sample app will need to have a push button that toggles drawing arrows between 5 points based on some "dateTime" property that's passed in from the host application. The arrows should identify which points have the earliest dateTime, order them earliest to latest and then draw an arrow from structure to structure. At this moment, the sample app does not need bounding box support, just work on sample data.

- [x] Define an appropriate dateTime property and testStatus property (positive, negative, not-tested) that we want to track on each structure "passed in by the host application"
- [x] Create 10 points of demo data with the dateTime property. Six are blue representing cases found  (testStatus = positive) and four are purple representing visited but no cases found (testStatus = negative).
- [x] Add the business logic that does the calculations (1. Find all points that have the dateTime property on the screen; 2. Within those, find all the points that have the testStatus = positive; 3. Order the points;
- [x] Add a button to the GeoWidget that draws the arrows. No animation is needed, we just need arrows to show up on the map

### LINKED ISSUE #259: G3.8 Filter Cases on the Map
The host application will need the ability to filter locations on the map based on business logic.

Pass Criteria:
- [ ] The host application can filter based on the type of location
- [ ] The host application can pass an array or list of locations to display and all other locations will be removed from the view

### PR DESCRIPTION
**Pull request by ekigamba**
Fixes #237
Fixes #259

### CODE REVIEW

File: library/src/main/java/io/ona/kujaku/exceptions/InvalidArrowLineConfig.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package io.ona.kujaku.exceptions;
+
+/**
+ * Created by Ephraim Kigamba - ekigamba@ona.io on 11/02/2019
+ */
+
+public class InvalidArrowLineConfig extends Exception {
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/exceptions/InvalidArrowLineConfig.java]: Could you rename this  `InvalidArrowLineConfigException` for better clarity?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java [hunk 2]
```diff
@@ -313,8 +312,143 @@ public void setLayerPropertiesFromBuilder() {
         assertEquals(propertyValues.get("line-width").value, lineWidth);
     }
 
+    @Test
+    public void constructorShouldThrowExceptionWhenDateTimeFormatIsNotSet() {
```
Comment (vincent-karuri) [inline library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: You can use  `@Test(expected = InvalidArrowLineConfig.class)` to simplify testing an exception is thrown as expected.

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java [hunk 3]
```diff
@@ -413,7 +411,168 @@ public void sortFeaturesShouldSortDescByDateTime() throws InvalidArrowLineConfig
...
+        assertEquals(5, sortedFeatures.size());
+
+        for (int i = 0; i < 5; i++) {
+            Feature sortedFeature = sortedFeatures.get(i);
+            assertEquals((int) sortedFeature.getNumberProperty("position"), i);
+        }
+    }
+
+    @Test
+    public void sortFeaturesShouldSortAscByString() throws InvalidArrowLineConfig {
```
Comment (vincent-karuri) [inline library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Generally, I don't think a test should throw an exception unless you expect it to.

Why not add a try-catch block?

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package io.ona.kujaku.comparisons;
+
+public abstract class Comparison {
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: I think this could be an interface instead?

There is no default functionality (concrete methods).

File: library/src/main/java/io/ona/kujaku/comparisons/EqualToComparison.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package io.ona.kujaku.comparisons;
+
+
+public class EqualToComparison extends Comparison {
+    private static final String TAG = "EqualToComparison";
+
+    @Override
+    public boolean compare(String a, String type, String b) {
+        switch (type) {
+            case TYPE_STRING:
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/comparisons/EqualToComparison.java]: Why does type need to be tested yet the arguments to be compared will always be type `String`?

File: library/src/main/java/io/ona/kujaku/utils/FeatureFilter.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+            if (filterConditions.size() > 0) {
+                for (Feature feature : featuresList) {
+                    boolean skipFeature = false;
+                    if (builder.getSortProperty() != null
+                            && !feature.hasProperty(builder.getSortProperty())) {
+                        continue;
+                    }
+
+                    for (FilterCondition filterCondition : filterConditions) {
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/utils/FeatureFilter.java]: I think you can define this`(filterCondition.getPropertyName()` once as it is used multiple times.

File: library/src/main/java/io/ona/kujaku/utils/FeatureFilter.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                        if (feature.hasProperty(filterCondition.getPropertyName())) {
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/utils/FeatureFilter.java]: I think you can define this`(filterCondition.getPropertyName()` once as it is used multiple times.

**Comment by vincent-karuri** [comment]
Not really necessary (required) but maybe you can add a toggle button that sorts the arrows alternately by `ASC` or `DESC`. To demonstrate the configurable sort functionality?

**Comment by vincent-karuri** [comment]
Actually according to the spec [here](https://github.com/onaio/kujaku/issues/237#issue-396650138), you might need to add a toggle button anyway.

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 2 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: It doesn't work with the current dependencies which end up using a method that is not available  or supported for the added dependencies.

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Exactly, so why catch it if it's not supposed to be thrown for an passing test

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java (hunk 4 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: The type's supported by the `Comparison` class which is declared as a variable and used later by child classes

File: library/src/main/java/io/ona/kujaku/comparisons/EqualToComparison.java (hunk 5 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/comparisons/EqualToComparison.java]: > Actually according to the spec [here](https://github.com/onaio/kujaku/issues/237#issue-396650138), you might need to add a toggle button anyway.

Cool, implemented according to the spec which says

> * Add a button to the GeoWidget that draws the arrows. No animation is needed, we just need arrows to show up on the map

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: I think the point was that the test should have correct inputs that should ensure it does not fail because of an invariant that is not under test.

The unit test should not fail because of something that the unit test is not testing.  In this case, it should only fail if the the sort did not happen as expected, like maybe wrong ordering. Not because the input was wrong.

The try-catch is simply because the originating function throws one not because we expect the test to throw one.

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package io.ona.kujaku.comparisons;
+
+public interface Comparison {
+    String TYPE_STRING = "string";
+
+    boolean compare(String a, String type, String b);
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: Do we need `String type` argument any more?

**Comment by vincent-karuri** [comment]
```
 The sample app will need to have a push button that toggles drawing arrows between 5 points based on some "dateTime" property that's passed in from the host application.
```

I think the button should toggle drawing arrows i.e. appear/disappear.

The disappear functionality should essentially be in the library and not in the sample activity?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: I find that if `InvalidArrowLineConfigException` were to extend `RuntimeException` or `IllegalArgumentException`, then you can use that syntax.

Do either of those work?

Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: https://stackoverflow.com/questions/2517965/what-exception-to-throw-when-an-important-parameter-dependency-is-missing

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java (hunk 8 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: I think it's good for us to maintain it for easy extensibility in case anyone else needs to add other Comparisons directly from [here](https://github.com/OpenSRP/opensrp-client-native-form/tree/master/android-json-form-wizard/src/main/java/com/vijay/jsonwizard/comparisons)

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 2 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: If you were to run the tests using that annoation, they will not run because the dependencies tend to mix up the versions which is currently not fixable. That is what I mean. You can use the annotation and run the tests yourself to see

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Right, the test should not fail. Why catch the exception or rather handle the exception if we don't expect it to fail? Do we want it to fail silently or continue passing even if an exception is thrown?

Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: It should not fail in the first place, I guess is what I'm trying to say i.e. if it is to fail, it should be for no other reason than the one under test.

The `try-catch` is not so as to handle the exception (the catch can be empty in this case) it's just because Java expects you to either wrap a method call that throws an exception in this block or in turn, throw the exception in the calling method i.e. it's just a formality to allow compiling.

The idea here would be that a developer is expected to know what are valid inputs for a method under test and provide those that wouldn't throw an exception where the exception is not under test.

The developer can then create a separate test to ensure that no exception is thrown for what they believe is valid input for a method under test. This method can have the `throws` signature if the developer chooses.

In short, let every test case test one thing only, and the method signature should clearly show what is under test. Adding a `throws exception` implies that an exception should be expected, when it shouldn't. Thoughts?

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java (hunk 8 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: Fair enough, and that's one way of doing it.

However, this approach delegates the responsibility of determining `type` to the calling function, right? 

Another approach may be letting the compare method impute the type from the `String a` and `String b` arguments thus having a central location for determining type and transferring that responsibility from the calling function(s)?

In this case, would we need the `type` argument?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: I had tried it out already and it works.

Please test and see, extend `IllegalArgumentException` in `InvalidArrowLineConfigException` and add `@Test(expected = InvalidArrowLineConfigException.class)`.

File: sample/src/main/java/io/ona/kujaku/sample/activities/CaseRelationshipActivity.java [hunk 9]
```diff
@@ -124,13 +123,18 @@ private void drawArrowsShowingRelationship() {
...
-                kujakuMapView.addArrowLineLayer(arrowLineLayer);
-                alreadyDrawn = true;
             } catch (InvalidArrowLineConfigException invalidArrowLineConfigException) {
                 Log.e(TAG, Log.getStackTraceString(invalidArrowLineConfigException));
             }
         }
+
+        if (arrowLineLayer.isVisible()) {
+            kujakuMapView.disableLayer(arrowLineLayer);
+            drawArrowsBtn.setText(R.string.disable_arrows_showing_relationship);
```
Comment (vincent-karuri) [inline sample/src/main/java/io/ona/kujaku/sample/activities/CaseRelationshipActivity.java]: I think you mixed up the text prompts?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Cool, so we add an empty `try-catch` and then if the developer breaks or changes the functionality, the test either passes(if we are helping it pass by catching the unexpected exception) or fails. If it fails, then it fails with no information on why it failed. The only information will be that it did not meet the assert condition. Is that easily debuggable? Do we even need to catch exceptions in tests if we don't expect the test to fail?  Even if the test is testing for something else?

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 2 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Yea, that works. Sorry, I meant the `ExpectedException` rule so that we can assert the message also.

Changing to `IllegalArgumentException` or any other child of `RuntimeException` is not ideal  for this situation since the app can still work OK without the `ArrowLineLayer` only that the layer will not be drawn. The error is also avoidable.

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java (hunk 8 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: > However, this approach delegates the responsibility of determining `type` to the calling function, right?

Yeap, this is OK by me, I want the developer to explicitly do this and the calling function to have the responsibility to do that. Just as in the native forms

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: OK.

File: library/src/main/java/io/ona/kujaku/comparisons/Comparison.java (hunk 8 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/comparisons/Comparison.java]: OK.

File: library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: I'm not sure I understand the issue being raised here.

Helper functions can be introduced in the tests to produce valid input. All tests needing valid input can call these methods. If the valid input is no longer valid due to code changes, the specific function that tests for valid input will fail, and you can then fix the helper function to produce the valid input.

Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: There is no need to make the test results more inaccurate and errors untraceable by catching errors which are not supposed to be thrown. The errors that are not supposed to be thrown are OK being thrown in the tests if the developer breaks or changes the code. `Try-catch`ing in a test does not help as it would in an app where it helps for the developer to recover from the error or rather fail silently without producing a bad UX. In this case, I want the tests to fail fast in case there is anything wrong even in preparation of the tests

Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/layers/ArrowLineLayerTest.java]: Let's agree on a difference in opinion.
